audio/data.py

```diff
@@ -258,66 +262,86 @@ def __init__(self, items, path, config=AudioConfig(), **kwargs):
+       
+        if self.config.use_spectro:
+            self.add_spectro(item)
+
+        return item
+
+    def _validate_consistencies(self, item):
+        if(self.config._sr is not None and item.sr != self.config._sr):
+            raise ValueError(f'''Multiple sample rates detected. Sample rate {item.sr} of file {fn} 
+                                does not match config sample rate {self.config._sr} 
                                 this means your dataset has multiple different sample rates, 
                                 please choose one and set resample_to to that value''')
-        if(sig.shape[0] > 1):
-            if not cfg.downmix:
-                warnings.warn(f'''Audio file {p} has {sig.shape[0]} channels, automatically downmixing to mono, 
-                                set AudioConfig.downmix=True to remove warnings''')
-            sig = DownmixMono(channels_first=True)(sig)
-        if cfg.max_to_pad or cfg.segment_size:
-            pad_len = cfg.max_to_pad if cfg.max_to_pad is not None else cfg.segment_size
-            sig = tfm_padtrim_signal(sig, int(pad_len/1000*sr), pad_mode="zeros")
-
-        mel = None
-        if cfg.use_spectro:
-            if cfg.mfcc: mel = MFCC(sr=sr, n_mfcc=cfg.sg_cfg.n_mfcc, melkwargs=cfg.sg_cfg.mel_args())(sig)
-            else:
-                mel = MelSpectrogram(**(cfg.sg_cfg.mel_args()))(sig)
-                if cfg.sg_cfg.to_db_scale: mel = SpectrogramToDB(top_db=cfg.sg_cfg.top_db)(mel)
-            mel = mel.permute(0, 2, 1)
-            if cfg.standardize: mel = standardize(mel)
-            if cfg.delta: mel = torch.stack([m.squeeze(0) for m in [mel, torchdelta(mel), torchdelta(mel, order=2)]]) 
-            if cfg.cache:
-                os.makedirs(image_path.parent, exist_ok=True)
-                torch.save(mel, image_path)
-                _record_cache_contents(cfg, [image_path])
-            start, end = None, None
-            if cfg.duration and cfg._processed: 
-                mel, start, end = tfm_crop_time(mel, cfg._sr, cfg.duration, cfg.sg_cfg.hop, cfg.pad_mode)
-        return AudioItem(sig=sig.squeeze(), sr=sr, spectro=mel, path=item, start=start, end=end)
+
+        if self.nchannels is None: self.nchannels = item.nchannels
+            
+        if item.nchannels > 1:
+            if self.config.downmix:
+                item.sig = DownmixMono(channels_first=True)(item.sig)
+
+            elif(self.nchannels != item.nchannels):
+                raise ValueError(f'''Multiple channel sizes detected. Channel size {item.nchannels} of file 
+                                {fn} does not match others' channel size of {self.nchannels}. A dataset may
+                                not contain different number of channels. Please set downmix=true in AudioConfig or 
+                                separate files with different number of channels.''')
+
+    def add_spectro(self, item: AudioItem):
```

Comment:
Really like the break down of all the logic into methods. Was something that was clearly needed :)